crypto/sm2/sm2_pmeth.c

```diff
@@ -0,0 +1,323 @@
+                              unsigned char *out, size_t *outlen,
+                              const unsigned char *in, size_t inlen)
+{
+    int ret;
+    int md_type;
+    EC_KEY *ec = ctx->pkey->pkey.ec;
+    SM2_PKEY_CTX *dctx = ctx->data;
+
+    if (dctx->md != NULL)
+        md_type = EVP_MD_type(dctx->md);
+    else
+        md_type = NID_sm3;
+
+    if (out == NULL) {
+        if (!sm2_ciphertext_size(ec, EVP_get_digestbynid(md_type), inlen,
+                                 outlen))
+            ret = -1;
+        else
+            ret = 1;
+    } else {
+        ret = sm2_encrypt(ec, EVP_get_digestbynid(md_type),
+                          in, inlen, out, outlen);
+    }
+
+    return ret;
+}
+
+static int pkey_sm2_decrypt(EVP_PKEY_CTX *ctx,
+                              unsigned char *out, size_t *outlen,
+                              const unsigned char *in, size_t inlen)
+{
+    int ret = -1;
+    int md_type;
+    EC_KEY *ec = ctx->pkey->pkey.ec;
+    SM2_PKEY_CTX *dctx = ctx->data;
+
+    if (dctx->md != NULL)
+        md_type = EVP_MD_type(dctx->md);
+    else
+        md_type = NID_sm3;
+
+    if (out == NULL) {
+        if (!sm2_plaintext_size(ec, EVP_get_digestbynid(md_type), inlen,
+                                outlen))
+            ret = -1;
+        else
+            ret = 1;
+    } else {
+        ret = sm2_decrypt(ec, EVP_get_digestbynid(md_type),
+                          in, inlen, out, outlen);
+    }
+
+    return ret;
+}
+
+static int pkey_sm2_ctrl(EVP_PKEY_CTX *ctx, int type, int p1, void *p2)
+{
+    SM2_PKEY_CTX *dctx = ctx->data;
+    EC_GROUP *group;
+
+    switch (type) {
+    case EVP_PKEY_CTRL_EC_PARAMGEN_CURVE_NID:
+        group = EC_GROUP_new_by_curve_name(p1);
+        if (group == NULL) {
+            SM2err(SM2_F_PKEY_SM2_CTRL, SM2_R_INVALID_CURVE);
+            return 0;
+        }
+        EC_GROUP_free(dctx->gen_group);
+        dctx->gen_group = group;
+        return 1;
+
+    case EVP_PKEY_CTRL_EC_PARAM_ENC:
+        if (dctx->gen_group != NULL) {
+            SM2err(SM2_F_PKEY_SM2_CTRL, SM2_R_NO_PARAMETERS_SET);
+            return 0;
+        }
+        EC_GROUP_set_asn1_flag(dctx->gen_group, p1);
+        return 1;
+
+    case EVP_PKEY_CTRL_MD:
+        if (EVP_MD_type((const EVP_MD *)p2) != NID_sha256
+            && EVP_MD_type((const EVP_MD *)p2) != NID_sha384
+            && EVP_MD_type((const EVP_MD *)p2) != NID_sha512
+            && EVP_MD_type((const EVP_MD *)p2) != NID_sm3) {
+            SM2err(SM2_F_PKEY_SM2_CTRL, SM2_R_INVALID_DIGEST_TYPE);
+            return 0;
+        }
+        dctx->md = p2;
+        return 1;
+
+    case EVP_PKEY_CTRL_GET_MD:
+        *(const EVP_MD **)p2 = dctx->md;
+        return 1;
+
+    default:
+        return -2;
+
+    }
+}
+
+static int pkey_sm2_ctrl_str(EVP_PKEY_CTX *ctx,
+                            const char *type, const char *value)
+{
+    if (strcmp(type, "ec_paramgen_curve") == 0) {
+        int nid;
```

Comment:
can merge the decl and the init code into one line (but don't forget the blank line)